### DISCUSSION TITLE
Enhance feature: better SAM / adetailer integration (mashb1t fork v2.5.0)

### DISCUSSION THREAD

**Discussion by mashb1t** [lllyasviel/Fooocus#3122] [Show and tell]
Now part of [v2.5.0-rc1](https://github.com/lllyasviel/Fooocus/releases/tag/v2.5.0-rc1) / https://github.com/lllyasviel/Fooocus/discussions/3282

~~WIP, but~~ works (had to fully restructure the async_worker to allow for multi-step processing).
You can still see the boxes, currently no feathering, just raw yolov8 and ultralytics. 

See https://github.com/mashb1t/Fooocus/compare/main...mashb1t:Fooocus:feature/multi-step-rendering

This also allows potentially automated generation + upscaling + adetailer in the future.

| generated image | merged mask | output |
|--------|--------|--------|
| ![image - 2024-06-08T233209 189](https://github.com/lllyasviel/Fooocus/assets/9307310/cdbe1d6a-ca1b-43db-a919-f3261fcb5de5)| ![image - 2024-06-08T234128 057](https://github.com/lllyasviel/Fooocus/assets/9307310/866c6f7e-b8dd-4153-ab3f-fce553b1d23f) | ![image - 2024-06-08T233210 543](https://github.com/lllyasviel/Fooocus/assets/9307310/60639cca-85c7-4f2f-b2e1-9ac33f93a5e5) | 

![screencapture-localhost-7865-2024-06-08-23_32_28](https://github.com/lllyasviel/Fooocus/assets/9307310/e961657f-660a-48c6-ac83-9a207b973cac)

**Comment by mashb1t** [comment]
Eyes model (https://civitai.com/models/178518/eyeful-or-robust-eye-detection-for-adetailer):

| Before | After |
|--------|--------|
| ![image - 2024-06-08T235752 807](https://github.com/lllyasviel/Fooocus/assets/9307310/736eb70d-55d1-4242-8417-0433a8bea8ea) | ![image - 2024-06-08T235754 509](https://github.com/lllyasviel/Fooocus/assets/9307310/6d991956-cdfc-49bc-a4c6-8cc7f75ccea4) |

**Comment by PeterJared** [comment]
Looks quite good, I find the current scaler a bit basic and this could improve a lot the results.

**Reply by mashb1t** [reply]
optimized with GroundingDINO + SAM, so masks can be prompted + are (most of the time) very good

| Yesterday (ultralytics/adetailer)| Today (GroundingDINO + SAM, no ultralytics/adetailer)|
|--------|--------|
| ![image](https://github.com/lllyasviel/Fooocus/assets/9307310/0aefdf74-82bb-4228-b1ba-590137eb9758) |![Screenshot 2024-06-09 225731](https://github.com/lllyasviel/Fooocus/assets/9307310/557f9081-f35f-45c5-a4cc-69f0bfb27136)|

**Comment by xhoxye** [comment]
Probably try to fix a broken human face

**Reply by mashb1t** [reply]
works, but a limitation is that in order to use an existing image (and you don't have the prompt anymore, as me in this example) you have to use inpaint and paint a small area not affecting the composition.
But this isn't the use case anyways (for now), the main intention is to auto-fix generated images.

<details><summary>UI</summary>
<p>

![screencapture-localhost-7865-2024-06-17-00_13_50](https://github.com/lllyasviel/Fooocus/assets/9307310/f2e6f9ac-87e0-46ca-b10e-ba6d67ab8724)

</p>
</details> 

 
| [Original](https://gallery.mashb1t.de/gallery/-8ubXvDC-HwkxOv8Oem8cabO/vd1Wm0E8WZdNbimHmVzTXTR_) | Variant 1 (0-shot) | Variant 2 (0-shot) |
|--------|--------|--------|
| ![22732ab04379cb552d97ebf38de6](https://github.com/lllyasviel/Fooocus/assets/9307310/06aced0d-a967-4b58-ac5a-ae7fd5fbf62a) | ![image - 2024-06-17T001256 976](https://github.com/lllyasviel/Fooocus/assets/9307310/8569db41-f6ae-44d6-8b83-1b937cdc51f4) | ![image - 2024-06-17T001252 624](https://github.com/lllyasviel/Fooocus/assets/9307310/33fd2d83-73c0-4192-b99e-515c39ae853f)| 

Both variants are from the same batch, only one enhancement iteration.

**Comment by mashb1t** [comment]
![image](https://github.com/lllyasviel/Fooocus/assets/9307310/0f2158b7-49cc-4365-9bc0-58e8673dd7d6)

remaining open ToDo's for UI:

- max num boxes in developer debug mode (currently hardcoded to 2)
- box_erode_or_dilate support
- steps (maybe same as upscaler?)
- detection + replacement prompt
- control for dino mask or sam mask (dino_debug) => use boxes and skip sam or use sam instead
-  (optional) inpaint mode select
- save intermediate stage2 images to developer debug mode
- save stage2 masks to developer debug mode
- display stage2 masks to developer debug mode
-  (optional) enabled tab => green

**Comment by mashb1t** [comment]
Quick update: still a few things on the ToDo list, but the technical base is done. The new and significantly better structured generation process also allows for positive and negative prompt adjustments as well as future expansion to upscale directly after generation (if needed) + to use different steps / dilate and erode values from the global setting per iteration.

Demo screenshots of branch [multi-step-rendering](https://github.com/mashb1t/Fooocus/tree/feature/multi-step-rendering) ([diff](https://github.com/mashb1t/Fooocus/compare/main...mashb1t:Fooocus:feature/multi-step-rendering)):

| without mask debugging | with mask debugging|
|--------|--------|
| ![screencapture-localhost-7865-2024-06-16-23_52_23](https://github.com/lllyasviel/Fooocus/assets/9307310/acb90904-989d-46b2-8fe9-4b1186ad95f5) | ![screencapture-localhost-7865-2024-06-16-23_53_56](https://github.com/lllyasviel/Fooocus/assets/9307310/1723be5d-54c0-4434-ae5d-0765867c8864) |

**Comment by PeterJared** [comment]
What are the plans for merging it?

**Reply by mashb1t** [reply]
> still a few things on the ToDo list

no fixed release date, but i still need to resolve the compatibility issue with image number >= 2, want to add a few minor UX improvements and testing has still to be done incl. all ControlNets.

I'll inform you when the feature is ready for testing.

**Reply by mashb1t** [reply]
@PeterJared code is 90%+ done, RC is dropping either tomorrow or on Friday

**Reply by PeterJared** [reply]
Cool, we have in YouTube already two guides for that feature, I guess outdated in some areas in this moment, specially the older one.

**Reply by mashb1t** [reply]
@PeterJared may i ask which videos you're referring to?

**Reply by PeterJared** [reply]
https://www.youtube.com/watch?v=RheMMNpWuzE&t=398s&ab_channel=KleebzTechAI

https://www.youtube.com/watch?v=9qnl8SnjRGA&t=191s&ab_channel=JumpIntoAI

**Reply by ILikeToasters** [reply]
I had a feeling my video was one of the ones you were talking about. Those cover just the auto masking feature in his fork and not this new implementation. I will probably have a new video once the feature is added and I figure it all out.

**Comment by mashb1t** [comment]
[v2.5.0-rc1](https://github.com/mashb1t/Fooocus/releases/tag/v2.5.0-rc1) is now available on my fork and I'm looking forward to your feedback.
 You can simply pull the tag for a fixed version or branch [feature/multi-step-rendering](https://github.com/mashb1t/Fooocus/tree/feature/multi-step-rendering) for the latest version.

Open ToDo's: https://github.com/mashb1t/Fooocus/discussions/40#discussioncomment-9811180

**Reply by ILikeToasters** [reply]
Will be checking it out very soon!

**Reply by mashb1t** [reply]
**HINT: use with performance Quality or Speed for best results when using an inpaint engine (inpaint mode default or add objects, not improve details), as the inpaint engine models are LoRAs and are not compatible with performance LoRAs on larger inpainting areas.** This is also true for normal inpainting, not only with the new enhance feature.

**Reply by ILikeToasters** [reply]
I love that you used "enhance".... "Enhance 224 to 176. Enhance, stop. Move in, stop. Pull out, track right, stop. Center in, pull back. Stop. Track 45 right. Stop. Center and stop. Enhance 34 to 36. Pan right and pull back. Stop. Enhance 34 to 46. Pull back. Wait a minute, go right, stop. Enhance 57 to 19. Track 45 left. Stop. Enhance 15 to 23. Give me a hard copy right there." - Rick Deckard

**Reply by DavidDragonsage** [reply]
Oooh Bladerunner, yes! 😎

@mashb1t  This is going to be great! Thank you for all the hard work! 🧙👍

**Comment by Alexdnk** [comment]
@mashb1t Why does the input field for "Detection prompt" only appear when model "sam" is selected? 
Are the models in some way specific and narrowly focused on something in particular? If so, could it be possible to show in a few words somewhere near the model selection what is the specialty of the currently selected model?
![1](https://github.com/lllyasviel/Fooocus/assets/83111151/77f73ed1-1693-4cee-897c-2be2b6b32059)

**Reply by mashb1t** [reply]
The other models do not support the usage of detection with prompts, only general subject masking. The model descriptions can be found here:
https://github.com/danielgatis/rembg?tab=readme-ov-file#models
Would you propose to add a link to Fooocus discussion with an explanatory page in a discussion, a separate readme file or some different solution?
(fyi i've implemented SAM from scratch for better performance and higher level of control, not using rembg)

**Reply by Alexdnk** [reply]
Thanks, it became clearer. The description in the future should be made in the style of the interface of Fooocus with such a link "📔 Document". Thank you for your hard work.

**Comment by mashb1t** [comment]
Upscale or Vary has been added as automated 1st option in enhance, closes https://github.com/lllyasviel/Fooocus/issues/3113

**Reply by ILikeToasters** [reply]
Getting an error with Upscale (Fast 2x). 

Fooocus] Processing enhance ...
[Fooocus] Downloading upscale models ...
[Fooocus] Upscaling image from (1024, 1024) ...
Upscaling image with shape (1024, 1024, 3) ...
Image upscaled.
[Fooocus] Saving image 1/2 to system ...
Traceback (most recent call last):
  File "F:\Mashb1t testing\Fooocus\modules\async_worker.py", line 1374, in worker
    handler(task)
  File "F:\Mashb1t testing\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\Mashb1t testing\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\Mashb1t testing\Fooocus\modules\async_worker.py", line 1274, in handler
    current_progress, img = process_enhance(
  File "F:\Mashb1t testing\Fooocus\modules\async_worker.py", line 976, in process_enhance
    uov_input_image_path = log(async_task.uov_input_image, d, output_format=async_task.output_format)
  File "F:\Mashb1t testing\Fooocus\modules\private_logger.py", line 31, in log
    image = Image.fromarray(img)
  File "F:\Mashb1t testing\python_embeded\lib\site-packages\PIL\Image.py", line 2943, in fromarray
    arr = obj.__array_interface__
AttributeError: 'NoneType' object has no attribute '__array_interface__'

**Reply by mashb1t** [reply]
@ILikeToasters i assume you're on the branch not on the tag, and you're using the step 1 upscaling and not the normal one, correct? I can check this issue tomorrow.

**Reply by ILikeToasters** [reply]
Yes on the `feature/multi-step-rendering` branch and using the enhance to auto upscale. The other upscales work fine just not the fast one it seems.

**Reply by mashb1t** [reply]
fixed, please pull again

**Comment by bobsponge76** [comment]
Hello, how can I try this fork? Should I just download it and overwrite the regular Fooocus installation?

**Reply by ILikeToasters** [reply]
I would suggest using a different folder so you can still use the original if you want. You can use the config to configure them to share the same model folders.

**Reply by mashb1t** [reply]
@bobsponge76 you can find the instructions in https://github.com/mashb1t/Fooocus?tab=readme-ov-file#switching-from-fooocus-to-this-fork.
To be 100% safe you can alternatively also follow @ILikeToasters variant of installing my fork in a new folder (using the [zip file](https://github.com/mashb1t/Fooocus/releases/download/2.1.864/Fooocus_mashb1t_win64_2-1-864.7z)) and then adjust the config + switch the branch there.

**Comment by mashb1t** [comment]
> works, but a limitation is that in order to use an existing image (and you don't have the prompt anymore, as me in this example) you have to use inpaint and paint a small area not affecting the composition.
But this isn't the use case anyways (for now), the main intention is to auto-fix generated images.

https://github.com/lllyasviel/Fooocus/discussions/3122#discussioncomment-9788700

Another update: it's now possible to start with an image without having to do inpaint/upscale before by using the new tab Input Image > Enhance.
@xhoxye fyi


`--enable-describe-uov-image` has also been renamed to `--enable-auto-describe-image` and the enhance image now also uses auto descriptions if the arg is set.

![screencapture-localhost-7865-2024-06-22-00_19_42](https://github.com/lllyasviel/Fooocus/assets/9307310/6833ece0-16fe-4419-9582-4dd7f10aaa92)

**Comment by mashb1t** [comment]
Draft Release Notes: https://github.com/mashb1t/Fooocus/discussions/42

Now with support for upscaling either before or after enhancement

**Comment by mashb1t** [comment]
v2.5.0 has been released in my fork, see https://github.com/mashb1t/Fooocus/releases/tag/v2.5.0. Happy to hear your opinion!

**Reply by ILikeToasters** [reply]
Would it be better to maybe change "Enable Mask Upload" to something different since it also enables the mask generation?

**Reply by mashb1t** [reply]
@ILikeToasters Hmm it does both, enable mask upload and show the mask generation functionality. What would you propose to change it to?

**Reply by Alexdnk** [reply]
I think mask generation model u2net_cloth_seg  is not working correctly. A tri-level mask is drawn. And it's superimposed exactly as it is, which is wrong. Along with this such an error in the console. How to repeat - on the screenshot.
```
Traceback (most recent call last):
  File "C:\Users\Alex\AppData\Local\Programs\Fooocus\Fooocus-mashb1t\Fooocus\modules\async_worker.py", line 1435, in worker
    handler(task)
  File "C:\Users\Alex\AppData\Local\Programs\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "C:\Users\Alex\AppData\Local\Programs\Fooocus\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "C:\Users\Alex\AppData\Local\Programs\Fooocus\Fooocus-mashb1t\Fooocus\modules\async_worker.py", line 1382, in handler
    current_progress, img, enhance_prompt_processed, enhance_negative_prompt_processed = process_enhance(
  File "C:\Users\Alex\AppData\Local\Programs\Fooocus\Fooocus-mashb1t\Fooocus\modules\async_worker.py", line 1013, in process_enhance
    imgs, img_paths, current_progress = process_task(all_steps, async_task, callback, controlnet_canny_path,
  File "C:\Users\Alex\AppData\Local\Programs\Fooocus\Fooocus-mashb1t\Fooocus\modules\async_worker.py", line 312, in process_task
    imgs = [inpaint_worker.current_task.post_process(x) for x in imgs]
  File "C:\Users\Alex\AppData\Local\Programs\Fooocus\Fooocus-mashb1t\Fooocus\modules\async_worker.py", line 312, in <listcomp>
    imgs = [inpaint_worker.current_task.post_process(x) for x in imgs]
  File "C:\Users\Alex\AppData\Local\Programs\Fooocus\Fooocus-mashb1t\Fooocus\modules\inpaint_worker.py", line 258, in post_process
    result[a:b, c:d] = content
ValueError: could not broadcast input array from shape (3588,832,3) into shape (1156,832,3)
```
![1lhkkz1i](https://github.com/lllyasviel/Fooocus/assets/83111151/989d3bde-03e8-49da-be91-d61df3876070)

**Reply by mashb1t** [reply]
@Alexdnk thanks for the hint, i'll check this in the evening

**Reply by mashb1t** [reply]
@Alexdnk fixed & pushed in main, see https://github.com/mashb1t/Fooocus/commit/b3a4b4e532eadaf545291f6db0553f25d8678bf1. Seems like i missed this field when refactoring the code. Thank you for testing & reporting, much appreciated!

**Reply by ILikeToasters** [reply]
> @ILikeToasters Hmm it does both, enable mask upload and show the mask generation functionality. What would you propose to change it to?

Maybe "Enable Advanced Masking Features" or "Enable Mask Upload/Auto-Gen"?

**Reply by mashb1t** [reply]
@ILikeToasters renamed to "Enable Advanced Masking Features" and pushed to main, thanks! Was thinking about "Manual / Automatic Masking" but one can already do manual masking without enabling the checkbox.

**Comment by Kaneki010** [comment]
i'm really hyped up for when this gets merged with the ability to automate it simply because of how time-consuming it is to inpaint everytime i get a bad face in a picture

**Comment by ZeroCool22** [comment]
Can someone do a tutorial video showcasing these new features¡

**Reply by mashb1t** [reply]
see https://github.com/mashb1t/Fooocus/discussions/42#discussioncomment-9897490, @ILikeToasters is https://www.youtube.com/@KLEEBZTECH

**Reply by ILikeToasters** [reply]
@ZeroCool22 This should get you started. https://youtu.be/NiGbksd9B_Y

**Comment by mashb1t** [comment]
Now part of [v2.5.0-rc1](https://github.com/lllyasviel/Fooocus/releases/tag/v2.5.0-rc1), docs in https://github.com/lllyasviel/Fooocus/discussions/3282
